examples_apps/run_state_learner.py

```diff
@@ -0,0 +1,432 @@
+#
+# By minimizing this cost function, the variational quantum circuit will
+# prepare a state with high fidelity to the target state.
+
+
+def cost(weights):
+    # Mapping from the Strawberry Fields free parameters to
+    # the TensorFlow variables.
+    mapping = {
+        p.name: w
+        for (sf_layer, weights_layer) in zip(sf_params, weights * 1.0)
+        for (p, w) in zip(sf_layer, weights_layer)
+    }
+
+    # Run engine
+    state = eng.run(prog, args=mapping).state
+
+    # Extract the statevector
+    ket = state.ket()
+
+    # Compute the fidelity between the output statevector
+    # and the target state.
+    fidelity = tf.abs(tf.reduce_sum(tf.math.conj(ket) * target_state)) ** 2
+
+    # Objective function to minimize
+    cost = tf.abs(tf.reduce_sum(tf.math.conj(ket) * target_state) - 1)
+    return cost, fidelity, ket
+
+
+######################################################################
+# Now that the cost function is defined, we can define and run the
+# optimization as well as the quantum engine. Below, we choose the Adam
+# optimizer that is built into TensorFlow, as well as instantiating
+# the Strawberry Fields TensorFlow backend.
+
+opt = tf.keras.optimizers.Adam(learning_rate=lr)
+eng = sf.Engine("tf", backend_options={"cutoff_dim": cutoff})
+
+######################################################################
+# We then loop over all repetitions, storing the best predicted fidelity
+# value.
+
+fid_progress = []
+best_fid = 0
+
+
+for i in range(reps):
+    # reset the engine if it has already been executed
+    if eng.run_progs:
+        eng.reset()
+
+    with tf.GradientTape() as tape:
+        loss, fid, ket = cost(weights)
+
+    # Stores fidelity at each step
+    fid_progress.append(fid.numpy())
+
+    if fid > best_fid:
+        # store the new best fidelity and best state
+        best_fid = fid.numpy()
+        learnt_state = ket.numpy()
+
+    # one repitition of the optimization
+    gradients = tape.gradient(loss, weights)
+    opt.apply_gradients(zip([gradients], [weights]))
+
+    # Prints progress at every rep
+    if i % 1 == 0:
+        print("Rep: {} Cost: {:.4f} Fidelity: {:.4f}".format(i, loss, fid))
+
+
+######################################################################
+# Results and visualisation
+# -------------------------
+#
+# Plotting the fidelity vs. optimization step:
+
+from matplotlib import pyplot as plt
+
+plt.rcParams["font.family"] = "serif"
+plt.rcParams["font.sans-serif"] = ["Computer Modern Roman"]
+plt.style.use("default")
+
+plt.plot(fid_progress)
+plt.ylabel("Fidelity")
+plt.xlabel("Step")
+
+
+######################################################################
+# We can use the following function to plot the Wigner function of our
+# target and learnt state:
+
+import matplotlib.pyplot as plt
+from mpl_toolkits.mplot3d import Axes3D
+
+
+def wigner(rho):
+    """This code is a modified version of the 'iterative' method
```

Comment:
An entry will have to be added to `LICENSE` on this file (and perhaps the license is needed at the beginning of this file as well? :thinking: ).

Reply:
> and perhaps the license is needed at the beginning of this file as well?

We don't typically worry about adding licenses to each file in the documentation, as there are a lot of files here that don't allow commenting! This is fine, since the license holds for the entire repository.

> An entry will have to be added to LICENSE on this file 

I wouldn't worry about this for now, since all tutorials are due to be moved to a separate repository soon.